proposals/extensions.md

```diff
@@ -399,54 +399,194 @@ These requirements need more refinement as implementation progresses, and may ne
+- erasing tuple names
+- erasing nullability annotations
+- erasing `notnull` constraints
+
+Note: other constraints are preserved, such as `new()`, `struct`, `class`, `allows ref struct`, `unmanaged`, and type constraints.
 
-Each extension declaration in source is emitted as an extension declaration in metadata (sometimes referred to as skeleton type).  
-- Its name is unspeakable and determined based on the lexical order in the program.  
-  The name is not guaranteed to remain stable across re-compilation. 
-  Below we use `<>E__` followed by an index. For example: `<>E__2`.  
-- Its type parameters are those declared in source (including attributes).  
+#### Extension grouping types
+
+Each extension block with a distinct CLR-level signature in source is emitted as an extension grouping type in metadata (extension grouping types are an evolution of the previous "skeleton type" design).  
+- Its name is unspeakable and determined based on the contents of the CLR-level signature. More details below.  
+- Its type parameters have normalized names (`T0`, `T1`, ...) and have no attributes.  
 - Its accessibility is public.
+- It is marked with the `specialname` flag and an `[Extension]` attribute.  
+
+The content-based name of the extension grouping type is based on the CLR-level signature and includes the following:
+- The fully qualified CLR name of the type. This will include namespaces, type constraints, named constraints (like `new`).
+    - Type parameter names will be normalized to `T0`, `T1`, etc ... based on the order they appear in the type declaration. This means an extension block with a type parameter `Dictionary<TKey, TValue>` will result in the fully qualified name being `System.Collections.Generic.Dictionary<T0, T1>`.
+    - The fully qualified name will not include the containing assembly. It is common for types to be moved between assemblies and that should not break the public API.
+- Constraints will included and sorted such that reordering them in source code does not change the name. Specifically:
+    - Type parameter constraints will be listed in declaration order. The constraints for the Nth type parameter will occur before the Nth+1 type parameter.
+    - Base type and interface constraints will be sorted by comparing the full names ordinally
+    - Non-type constraints will be sorted by comparing the C# text ordinally
+- Since this does not include attributes, it intentionally ignores C#-isms like tuple names, nullability, etc ...
+
+Note: The name is guaranteed to remain stable across re-compilations, re-orderings and changes of C#-isms (ie. that don't affect the CLR-level signature).   
+
+#### Extension marker types
+
+The marker type re-declares the type parameters of its containing type (an extension grouping type) to get full fidelity of the C# view of extension blocks.
+
+Each extension block with a distinct C#-level signature in source results in an extension marker type in metadata.  
```

Comment:
Similar feedback as for grouping type. Consider saying something like: "An extension marker type is emitted to metadata for each set of extension blocks in source with a the same C#-level signature."